### PR TITLE
Ct 1223 clarisse offline path transform

### PR DESCRIPTION
**Pull request by hoolymama**
Hi, This PR is all Clarisse related. Also generic gpath stuff was changed, but that's not being used outside Clarisse either at the moment.

The purpose of this work is to change behavior of windows path resolution.
Previously, paths were "Linuxified" within Clarisse using it's API by way of a prerender script.
The problem was that connections between different referenced files (for example - rocks as input to a scatterer from a referenced rock library) would be broken on file open. Fixing the referenced paths after the scene opens does not reconnect. Therefore, the scene needs to have all valid path references before its opened, and for this reason, we write out linuxified copies of the scene file and all project files it references. The prerender script is still used for validations and so on, but it no longer manipulates paths.

### CODE REVIEW

File: conductor/clarisse/scripted_class/preview_ui.py [hunk 1]
```diff
@@ -149,6 +149,8 @@ def on_write_but(self, sender, eventid):
         """
         with cu.waiting_cursor():
             self.submission.write_render_package()
+            if os.name == "nt":
```
Comment (flebel) [inline conductor/clarisse/scripted_class/preview_ui.py]: Can you extract this into a helper method `is_windows()`, and re-use it where we check for `os.name == "nt"`?

File: conductor/clarisse/scripted_class/submission.py [hunk 2]
```diff
@@ -48,6 +55,37 @@
...
+
+    As we scan project files to replace windows paths, we use this regex which
+    will be something like: r'\s+(?:filename|filename_sys|save_as)\s+"(.*)"\s+' 
+    only longer.
+    """
+    classes = ix.application.get_factory().get_classes()
+    file_attrs = []
+    for klass in classes.get_classes():
+        attr_count = klass.get_attribute_count()
+        for i in range(attr_count):
```
Comment (flebel) [inline conductor/clarisse/scripted_class/submission.py]: Use `xrange` to return a generator.

File: conductor/clarisse/scripted_class/submission.py [hunk 3]
```diff
@@ -48,6 +55,37 @@
...
+            attr = klass.get_attribute(i)
+            hint = attr.get_visual_hint()
+            if hint in [
```
Comment (flebel) [inline conductor/clarisse/scripted_class/submission.py]: Extract the hardcoded list into a tuple outside of the loops.

File: conductor/clarisse/scripted_class/submission.py [hunk 4]
```diff
@@ -48,6 +55,37 @@
...
+                ix.api.OfAttr.VISUAL_HINT_FILENAME_SAVE,
+                ix.api.OfAttr.VISUAL_HINT_FILENAME_OPEN,
+                ix.api.OfAttr.VISUAL_HINT_FOLDER,
+            ]:
+                file_attrs.append(attr.get_name())
+
+    file_attrs = list(set(file_attrs))
```
Comment (flebel) [inline conductor/clarisse/scripted_class/submission.py]: `file_attrs = sorted(set(file_attrs))`

File: conductor/clarisse/scripted_class/submission.py [hunk 5]
```diff
@@ -319,6 +372,92 @@ def write_render_package(self):
...
+        os.remove(self.render_package_path.posix_path())
+        self._linuxify_file(
+            temp_path, PATH_LINE_REGEX, self.render_package_path.posix_path()
+        )
+
+    def _linuxify_file(self, filename, path_regex, dest_path=None):
+        """
+        Fix paths for one file.
+
+        If the file already has the .ct.project extension, replace it too,
```
Comment (flebel) [inline conductor/clarisse/scripted_class/submission.py]: `s/,$/.`/

File: conductor/clarisse/scripted_class/preview_ui.py (hunk 1 above)
Comment (hoolymama) [reply conductor/clarisse/scripted_class/preview_ui.py]: done

File: conductor/clarisse/scripted_class/submission.py (hunk 2 above)
Comment (hoolymama) [reply conductor/clarisse/scripted_class/submission.py]: done

File: conductor/clarisse/scripted_class/submission.py (hunk 3 above)
Comment (hoolymama) [reply conductor/clarisse/scripted_class/submission.py]: done

File: conductor/clarisse/scripted_class/submission.py (hunk 4 above)
Comment (hoolymama) [reply conductor/clarisse/scripted_class/submission.py]: done

File: conductor/clarisse/scripted_class/submission.py (hunk 5 above)
Comment (hoolymama) [reply conductor/clarisse/scripted_class/submission.py]: done

**Pull request merged**